build/get-webpack-config.js

```diff
@@ -465,17 +470,20 @@ function getWebpackConfig(opts /*: WebpackConfigOpts */) {
       // in dev because the CLI will not exit with an error code if the option is enabled,
       // so failed builds would look like successful ones.
       watch && new webpack.NoEmitOnErrorsPlugin(),
-      new InstrumentedImportDependencyTemplatePlugin(
-        runtime !== 'client'
-          ? // Server
+      runtime === 'server'
+        ? // Server
+          new InstrumentedImportDependencyTemplatePlugin(
             state.mergedClientChunkMetadata
-          : /**
-             * Client
-             * Don't wait for the client manifest on the client.
-             * The underlying plugin handles client instrumentation on its own.
-             */
-            void 0
-      ),
+          )
+        : /**
+           * Client
+           * Don't wait for the client manifest on the client.
+           * The underlying plugin is able determine client chunk metadata on its own.
+           */
+          new InstrumentedImportDependencyTemplatePlugin(
+            void 0,
```

Comment:
Now that this constructor now takes more than one parameter, I think an object might make this more understandable. This was already a bit confusing where the argument is expected to be undefined in the client case, but server-side should be provided. Adding another parameter now makes it even worse.

Expressing the parameter type as a disjoint union of two different objects would help I think. For example:
```js
type Opts = 
  | ClientOpts
  | ServerOpts;

type serverOpts = {
  clientChunkMetadata: ClientChunkMetadataState
};

type clientOpts = {
  i18nManifest: TranslationsManifest
};
```

Also, this makes me think: is it odd that `__I18N_KEYS` is only added on the client-side? The other two properties are added in both the server and client bundles.

Reply:
Yeah I think that'll make the plugin easier to understand. I'll update to use those types.

Reply:
I was actually thinking the opposite for the promise instrumentation. I don't totally understand why `__CHUNK_IDS` are added to the server build since they're only used when dynamically loading chunks with new translations (at least I think that's the only time they're used).

Reply:
`__CHUNK_IDS` is necessary server-side because that's how the server knows which async chunks are used in the SSR and are thus considered "critical" and should therefore be preloaded. During SSR, when it encounters a `split` component, it adds the chunk ids to the list of critical chunks that should be preloaded.

Reply:
Okay I recall that now. Since I didn't change how the translation manifest is added to the server bundle, the server still looks up the necessary translations based on the `__CHUNK_IDS` of the promise. We could potentially change that to use use `__I18N_KEYS` but yes probably not really necessary at this point. It might require a different approach than what I have because this approach doesn't find any translation keys since they're all bundled together in the server build.